nexus/db-queries/src/db/datastore/sled.rs

```diff
@@ -43,11 +48,244 @@ use omicron_uuid_kinds::GenericUuid;
+// - Preferred := Preferred ∖ Both
+// - Unpreferred := Unpreferred ∖ Both
+// - If Preferred isn't empty, pick a target from it.
+// - Targets := Targets \ Unpreferred
+// - If Targets isn't empty, pick a target from it.
+// - If Unpreferred isn't empty, pick a target from it.
+// - Fail, no targets are available.
+fn pick_sled_reservation_target(
+    log: &Logger,
+    mut targets: HashSet<SledUuid>,
+    anti_affinity_sleds: Vec<(AffinityPolicy, Uuid)>,
+    affinity_sleds: Vec<(AffinityPolicy, Uuid)>,
+) -> Result<SledUuid, SledReservationError> {
+    let (banned, mut unpreferred): (HashSet<_>, HashSet<_>) =
+        anti_affinity_sleds.into_iter().partition_map(|(policy, id)| {
+            let id = SledUuid::from_untyped_uuid(id);
+            match policy {
+                AffinityPolicy::Fail => Either::Left(id),
+                AffinityPolicy::Allow => Either::Right(id),
+            }
+        });
+    let (required, mut preferred): (HashSet<_>, HashSet<_>) =
+        affinity_sleds.into_iter().partition_map(|(policy, id)| {
+            let id = SledUuid::from_untyped_uuid(id);
+            match policy {
+                AffinityPolicy::Fail => Either::Left(id),
+                AffinityPolicy::Allow => Either::Right(id),
+            }
+        });
+
+    if !banned.is_empty() {
+        info!(
+            log,
+            "anti-affinity policy prohibits placement on {} sleds", banned.len();
+            "banned" => ?banned,
+        );
+    }
+    if !required.is_empty() {
+        info!(
+            log,
+            "affinity policy requires placement on {} sleds", required.len();
+            "required" => ?required,
+        );
+    }
+
+    if required.len() > 1 {
+        return Err(SledReservationError::TooManyAffinityConstraints);
+    }
+    if let Some(required_id) = required.iter().next() {
+        // If we have a "required" sled, it must be chosen.
+        if banned.contains(&required_id) {
+            return Err(
+                SledReservationError::ConflictingAntiAndAffinityConstraints,
+            );
+        }
+        if !targets.contains(&required_id) {
+            return Err(SledReservationError::RequiredAffinitySledNotValid);
+        }
+        return Ok(*required_id);
+    }
+
+    // We have no "required" sleds, but might have preferences.
+    targets = targets.difference(&banned).cloned().collect();
+
+    // Only consider "preferred" sleds that are viable targets
+    preferred = targets.intersection(&preferred).cloned().collect();
+    // Only consider "unpreferred" sleds that are viable targets
+    unpreferred = targets.intersection(&unpreferred).cloned().collect();
+
+    // If a target is both preferred and unpreferred, it is removed
+    // from both sets.
+    let both = preferred.intersection(&unpreferred).cloned().collect();
+    preferred = preferred.difference(&both).cloned().collect();
+    unpreferred = unpreferred.difference(&both).cloned().collect();
```

Comment:
the more i look at this, the more i wish that `HashSet` had an operation to turn itself into its intersection or difference with another set in place, reusing the prior allocation (so that rather than assigning to the new `targets`/`preferred`/`unpreferred`, we could just mutate the existing hashsets. probably not too important to allocation golf this today, but it really makes me wish such an API existed...